### PR TITLE
Add memory traits for dynamic memory using TLSF

### LINKED ISSUE #136: TLSF allocator breaks program
Hi!

I updated xpcc to the latest version recently for the I2C fix, but unfortunately my program stopped working properly, nothing is drawn on the screen nor do input buttons work. I used git bisect to find that the breaking commit is 0b457d65525a9ad8f50bd99d8ec961e8ae807c9b, which enables the TLSF allocator, but I don't use dynamic allocation at all. I didn't get to debugging yet (It's just so cumbersome to set up on windows), but from experimenting I suspect something in xpcc, more specifically in formatting (probably floats) might be the cause.

I'll comment when I know more.

### PR DESCRIPTION
**Pull request by salkinium**
This adds the concept of memory traits to the allocators, incl. global overloading of operator `new`.
This expands the work in #115 by allowing the user to request memory with certain traits, like DMA-able, Fast access, or external RAM.

All heap arenas are now stored in an array constructed in the linker script, which is used to initialize all allocators at runtime. If TLSF is not the chosen allocator, then newlib is initialized with the largest continous heap arena, that satifies the `xpcc::MemoryDefault` trait (ie. at least DMA-able and S-Bus access).

Memory traits can be requested using the overloaded `new` syntax:

``` cpp
Foo *foo = new (xpcc::MemoryFastData) Foo;
uint16_t *array = new (xpcc::MemoryExternal | xpcc::MemoryDMA) uint16_t[200];
```

There is a generic fallback mechanism, where the type of memory (External, Fast, but _not_ non-volatile!) is discarded when the memory is exhausted for in the primary pool and a second, relaxed search is done. Note that only the type is discarded, not the access method. So a request for DMA-able memory will not fallback to non-DMA-able memory.

Furthermore, the memory operations are now wrapped by the linker, which should fix #136, by properly guiding malloc to xpcc's implementation and not to newlib's, when it is not used.

Note that the memory traits are ignored for AVRs, on Hosted and when using the newlib or block allocator.
(Memory traits for AVRs are not impossible however,  as some AVRs have external memory interfaces.)

For newlib and block_allocator, the largest continuous heap with the `xpcc::MemoryDefault` trait (DMA-able and at least S-Bus access) is chosen from the heap table.
This means the heap location is not known at link time anymore, but is computed at boot time, potentially altering behavior. Practically this means that SRAM{1,2,3} are chosen as before, since they are the largest continuous memory region.
However, for the F469 with the external 16MB SDRAM, it is chosen as the primary heap.

~~This is an issue for the block allocator, which has 8B chunks, but only 2^16 of them (`uint16_t`).
Apparently it has an overflow condition where when passing regions larger than _8B \* 2^16 = 512kB_ creates an overflow with an infinite loop. Setting the chunk size to 256B (_256B \* 2^16 = 16MB_) fixes this. But this is obviously not a good design. (Who uses that block allocator anyway??)~~ (fixed)

cc @ekiwi @dergraaf @daniel-k

### CODE REVIEW

**Comment by salkinium** [comment]
This is missing the ignoring of the memory traits for the `hosted` target!

**Comment by salkinium** [comment]
I seem to have broken the linker, but I don't know why. Just adding backup SRAM into the linkerscript "fixes"it. Very weird.

**Comment by ekiwi** [comment]
This seems like a cool design that however forces us to switch to TLSF for it to be actually useful. However, there seem to be some problems with our new allocator (see issue #136), so are we really ready to switch allocators?

I know you want this change to be able to use the external SDRAM. As the main application seems to be a display buffer, couldn't you just create a static allocatioin on the SDRAM and place the heap in internal RAM?

**Comment by salkinium** [comment]
> However, there seem to be some problems with our new allocator (see issue #136), so are we really ready to switch allocators?

The issues aren't with TLSF, but with overwriting `malloc` instead of `_malloc_r`. Whenever newlib allocates memory it uses `_malloc_r` (directly, or indirectly), and this is now forcefully overwritten using linker wrapping.
The issues I have with TLSF is the 1kB overhead _per allocator_ (= per distinct memory trait). I'd be ok with 1kB overhead once, but that requires hacking TLSF to make it memory trait aware.

> I know you want this change to be able to use the external SDRAM. As the main application seems to be a display buffer, couldn't you just create a static allocatioin on the SDRAM and place the heap in internal RAM?

No, let's do this properly. The display requires an event loop for its events and that requires dynamic memory to be sane to program with it.
It's ill-advised to avoid dynamic memory operations for the sake of efficiency when compared to the graphic operations that take magnitudes more computational power than TLSF or newlib malloc.

Note that this change does not force you to use TLSF. It is merely one backend that allows the usage of all memory. The may also be another backend implementation like [umm_malloc](https://github.com/rhempel/umm_malloc).
The changes introduced in https://github.com/roboterclubaachen/xpcc/pull/147/commits/288dfb32fb667475ff87101504f24c16e0bf92f8 will make the newlib allocator use the 16MB external RAM and just ignore all ~300kB of internal RAM.

File: src/xpcc/architecture/platform/driver/core/cortex/syscalls.cpp.in [hunk 1]
```diff
@@ -140,25 +201,47 @@ _isatty(int /*file*/)
...
+	if (__brkval >= heap_end) {
 		// ERROR: heap and stack collision!
 		abort();
 	}
 
 	//  Return pointer to start of new heap area.
 	return heap;
 }
+
+// FIXME: "Unwrapping" the malloc for newlib
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/core/cortex/syscalls.cpp.in]: This could potentially be made nicer, by not setting the linker flags.
But that would require the linker to know about the parameters, which is not a mechanism that we have at the moment.

**Comment by salkinium** [comment]
Fixed the linker issue. I accidentally referenced `__heap5_start` and `__heap5_end` from the the non-existant `.backup` section in the `.table.heap`. The linker should have given me a line number instead of just the "invalid data statement" error 😩

**Comment by salkinium** [comment]
- Added the overloaded operator `new` implementation for hosted.
- Rebased all history to move each allocator implementation into its own file.
- Capped the block allocator size so that it does not break when too much memory is added.
- Made block allocator configurable via overrideable macro defaults.

File: src/xpcc/architecture/platform/driver/core/cortex/stm32/stm32_idtcm.ld.in [hunk 2]
```diff
@@ -486,6 +486,41 @@ SECTIONS
...
+		LONG (0x001f) /* SRAM1 */
+		LONG (__heap1_start)
+		LONG (__heap1_end)
+		LONG (0x001f) /* SRAM2 */
+		LONG (__heap2_start)
+		LONG (__heap2_end)
+		LONG (0x200f) /* DTCM */
+		LONG (__heap0_start)
+		LONG (__heap0_end)
+		LONG (0x2004) /* ITCM */
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/core/cortex/stm32/stm32_idtcm.ld.in]: Confirm that the ITCM cannot be accessed via D-Bus or similar.
How would you get the instruction data in there in the first place?
If it were not accessible by D-Bus, then the allocator pool initialization in this region will fail.

File: src/xpcc/architecture/platform/driver/core/cortex/driver.xml [hunk 3]
```diff
@@ -39,11 +39,21 @@
...
-		<template>syscalls.cpp.in</template>
+		<!-- dealing with C++ -->
 		<static>cxxabi.cpp</static>
+		<!-- dealing with newlib -->
+		<static>syscalls.c</static>
+		<!-- startup script -->
+		<template>startup.c.in</template>
 		<static>default_handler.sx</static>
+		<!-- everything to do with dynamic memory -->
+		<template>heap_newlib.c.in</template>
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/core/cortex/driver.xml]: Is it possible to filter for parameter settings?
Something like:
```xml
<template parameter-allocator="newlib">heap_newlib.c.in</template>
```

Maybe this should be a python file. XML seems the wrong format. cc @ekiwi

**Comment by salkinium** [comment]
Check usage of `exit(-1)` in newlib allocator on memory exhaustion, which leads to a system halt via infinite loop. Neither TLSF nor Block Allocator exhibit this behavior.
Since there is no good answer to memory exhaustion, I would recommend using `exit(-1)` on TLSF too, and in the `exit()` function to toggle the hardfault handler LED if set and to output a failure string via hardfault handler log uart.

(Done.)

File: src/xpcc/architecture/platform/driver/core/cortex/driver.xml [hunk 4]
```diff
@@ -38,11 +39,22 @@
...
+		<!-- dealing with C++ -->
 		<static>cxxabi.cpp</static>
+		<!-- dealing with newlib -->
+		<static>syscalls.c</static>
+		<!-- startup script -->
+		<template>startup.c.in</template>
 		<static>default_handler.sx</static>
+		<!-- everything to do with dynamic memory -->
+		<static>heap_table.c</static>
+		<template>heap_newlib.c.in</template>
```
Comment (salkinium) [inline src/xpcc/architecture/platform/driver/core/cortex/driver.xml:51]: Is it possible to filter for parameter settings?
Something like:

``` xml
<template parameter-allocator="newlib">heap_newlib.c.in</template>
```

Maybe this should be a python file. XML seems the wrong format. cc @ekiwi

Comment (ekiwi) [inline src/xpcc/architecture/platform/driver/core/cortex/driver.xml:51]: > Maybe this should be a python file. XML seems the wrong format

The problem with python is, that it is imho too powerful for configuration files.
I would suggest you place the allocator specific files in a new driver directory. Something like `dynamic_memory` or just `allocator` as the driver type.

Comment (salkinium) [inline src/xpcc/architecture/platform/driver/core/cortex/driver.xml:51]: > The problem with python is, that it is imho too powerful for configuration files.

Yes, I would like to agree, but it seems like we have written a very simple package manager that states the dependencies in the device files and the packages in our drivers.
@dergraaf's [library-builder](https://github.com/dergraaf/library-builder) uses Python as well, especially for the module options. So does [Conan](http://docs.conan.io/en/latest/conanfile_py.html#using-options).

There is an argument to be made for a more powerful language. For example the linkerscripts should actually be decided by the _type_ of memory available instead of switching on the device name:

``` xml
<template device-family="f4" device-name="405|407|415|417|427|429|437|439|469|479" out="linkerscript.ld">stm32/stm32_dccm.ld.in</template>
<template device-family="f4" device-name="401|410|411|446" out="linkerscript.ld">stm32/stm32_ram.ld.in</template>
```

The F401 etc do not have Core Coupled Memory, but I cannot query this in our language, since it is too restricted.
Similarly, not having an `else` in our XML language is a bit annoying in the above example.

I would propose leaving the implementation as it is right now, due to the scope of this PR, and investigate this later.
It would make sense to me to split the `driver/core/cortex` into `driver/core/cortex` (startup, linker, delay), `driver/libc/newlib` (heap, printf) and `driver/libc/avrlibc` (heap) or similar.
xpcc is still quite bad at integrating with newlib, and I would like to change that in the future.

File: src/xpcc/architecture/interface/memory.hpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+XPCC_TYPE_FLAGS(MemoryTraits);
+
+/// Fast code memory is accessible by at least the I-Bus and it is core-coupled.
+/// Fallback memory on exhaustion is not core-coupled.
+/// @ingroup memory_traits
+constexpr MemoryTraits MemoryFastCode = (MemoryTrait::AccessIBus | MemoryTrait::TypeCoreCoupled);
+/// Fast data memory is accessible by at least the D-Bus and it is core coupled.
+/// Fallback memory on exhaustion is not core-coupled.
+/// @ingroup memory_traits
+constexpr MemoryTraits MemoryFastData = (MemoryTrait::AccessDBus | MemoryTrait::TypeCoreCoupled);
```
Comment (salkinium) [inline src/xpcc/architecture/interface/memory.hpp:61]:
``` diff
- constexpr MemoryTraits MemoryFastData = (MemoryTrait::AccessDBus | MemoryTrait::TypeCoreCoupled);
+ constexpr MemoryTraits MemoryFastData = (MemoryTrait::AccessDBus | MemoryTrait::AccessSBus | MemoryTrait::TypeCoreCoupled);
```

File: src/xpcc/architecture/interface/memory.hpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+/// A memory can have multiple traits.
+/// @ingroup memory_traits
+using MemoryTraits = Flags16<MemoryTrait>;
+XPCC_TYPE_FLAGS(MemoryTraits);
+
+/// Fast code memory is accessible by at least the I-Bus and it is core-coupled.
+/// Fallback memory on exhaustion is not core-coupled.
+/// @ingroup memory_traits
+constexpr MemoryTraits MemoryFastCode = (MemoryTrait::AccessIBus | MemoryTrait::TypeCoreCoupled);
```
Comment (salkinium) [inline src/xpcc/architecture/interface/memory.hpp]:
```diff
- constexpr MemoryTraits MemoryFastData = (MemoryTrait::AccessIBus | MemoryTrait::TypeCoreCoupled);
+ constexpr MemoryTraits MemoryFastData = (MemoryTrait::AccessIBus | MemoryTrait::AccessSBus | MemoryTrait::TypeCoreCoupled);
```

File: src/xpcc/architecture/interface/memory.hpp (hunk 5 above)
Comment (salkinium) [inline src/xpcc/architecture/interface/memory.hpp:61]: Nope, this won't make it match the F4 internal CCM, which only has DBus access not SBus access.

Comment (salkinium) [inline src/xpcc/architecture/interface/memory.hpp:61]: A better understanding of the fallback mechanism is required. SBus can service both data and instruction requests, but slower, therefore DBus and IBus should always be preferred.

The current mask is interpreted as the _minimum_ requirements, ie. if we add `SBus` the CCM won't be matched.
On heap exhaustion there should perhaps be a three-step process: `requirement` -> `requirement & ~type` -> `(requirement & ~type) | SBus`?
However, that would incur searching three times in case of complete heap exhaustion.
On the other hand, searching longer is better than an out-of-memory scenario.

File: src/xpcc/architecture/platform/driver/core/cortex/driver.xml (hunk 4 above)
Comment (ekiwi) [inline src/xpcc/architecture/platform/driver/core/cortex/driver.xml:51]: > @dergraaf's library-builder uses Python as well, especially for the module options. So does Conan.

The problem with using fully fledged python is that IDE and other tool support is quite difficult. While CMake and make have reasonable IDE support, there is no such thing for SCons, imho because it is very hard to parse all dependencies.

Comment (salkinium) [inline src/xpcc/architecture/platform/driver/core/cortex/driver.xml:51]: Ok, I'm really not opposed to using CMake, I did integrate Jinja2 templating into CMake before and it wasn't totally terrible.

I'm not sure if we need IDE support for the HAL templates though. As @dergraaf pointed out some time ago, the HAL generation should really be a separate step and the resulting HAL should then be build together with the rest of the framework.
I would therefore favour using Python for the HAL generation part, with a CMake wrapper of course. Library builder is probably the better choice than Conan, due to being written with exactly this in mind.

File: src/xpcc/architecture/interface/memory.hpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+namespace xpcc
+{
+
+/// Describes access and type of dynamic memory.
+/// @ingroup memory_traits
+enum class
+MemoryTrait : uint16_t
+{
+	AccessSBus = Bit0,			///< Memory is accessible via System-Bus
```
Comment (salkinium) [inline src/xpcc/architecture/interface/memory.hpp:35]: At the moment multiple files need to be synchronized:
- this header,
- 3 heap implementations (TLSF, newlib and block allocator)
- 4 linkerscripts for stm32

Not aware of any way to introduce source definitions into the linkerfile though.

**Comment by salkinium** [comment]
@salkinium will you please include `<errno.h>` and use `exit(ENOMEM)` like a normal person.

(Done.)

File: src/xpcc/architecture/interface/memory.hpp (hunk 5 above)
Comment (salkinium) [inline src/xpcc/architecture/interface/memory.hpp:61]: [Implemented](https://github.com/roboterclubaachen/xpcc/pull/147/files#diff-880ea9bd10bea4f563083a670c9044adR110).
